grbl/stepper.c

```diff
@@ -351,40 +365,44 @@ ISR(TIMER1_COMPA_vect)
   st.step_outbits = 0; 
 
   // Execute step displacement profile by Bresenham line algorithm
-  #ifdef ADAPTIVE_MULTI_AXIS_STEP_SMOOTHING
-    st.counter_x += st.steps[X_AXIS];
-  #else
-    st.counter_x += st.exec_block->steps[X_AXIS];
-  #endif  
-  if (st.counter_x > st.exec_block->step_event_count) {
+  st.counter[X_AXIS] += st.steps[X_AXIS];
+
+  if (st.counter[X_AXIS] > st.exec_block->step_event_count) {
     st.step_outbits |= (1<<X_STEP_BIT);
-    st.counter_x -= st.exec_block->step_event_count;
+    st.counter[X_AXIS] -= st.exec_block->step_event_count;
     if (st.exec_block->direction_bits & (1<<X_DIRECTION_BIT)) { sys.position[X_AXIS]--; }
     else { sys.position[X_AXIS]++; }
   }
-  #ifdef ADAPTIVE_MULTI_AXIS_STEP_SMOOTHING
-    st.counter_y += st.steps[Y_AXIS];
-  #else
-    st.counter_y += st.exec_block->steps[Y_AXIS];
-  #endif    
-  if (st.counter_y > st.exec_block->step_event_count) {
+  
+  st.counter[Y_AXIS] += st.steps[Y_AXIS];
+
+  if (st.counter[Y_AXIS] > st.exec_block->step_event_count) {
     st.step_outbits |= (1<<Y_STEP_BIT);
-    st.counter_y -= st.exec_block->step_event_count;
+    st.counter[Y_AXIS] -= st.exec_block->step_event_count;
     if (st.exec_block->direction_bits & (1<<Y_DIRECTION_BIT)) { sys.position[Y_AXIS]--; }
     else { sys.position[Y_AXIS]++; }
   }
-  #ifdef ADAPTIVE_MULTI_AXIS_STEP_SMOOTHING
-    st.counter_z += st.steps[Z_AXIS];
-  #else
-    st.counter_z += st.exec_block->steps[Z_AXIS];
-  #endif  
-  if (st.counter_z > st.exec_block->step_event_count) {
+  
+  st.counter[Z_AXIS] += st.steps[Z_AXIS];
+
+  if (st.counter[Z_AXIS] > st.exec_block->step_event_count) {
     st.step_outbits |= (1<<Z_STEP_BIT);
-    st.counter_z -= st.exec_block->step_event_count;
+    st.counter[Z_AXIS] -= st.exec_block->step_event_count;
```

Comment:
I did not check the assembly.  Making an array allowed cleaner looking c
code than the alternative with multiple #ifdefs.     But I agree
performance overrules style considerations. I will take a look later.